### PR TITLE
plugginable arraySize

### PR DESCRIPTION
**Pull request by izulin**
### Context

Introduces plugginable arraySize checks.

WIP (tests missing, docs missing, exports missing)

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #760 
2.
3.

### Checklist:

- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation,
- [x] I described the modification in the CHANGELOG.md file.

### CODE REVIEW

**Comment by izulin** [comment]
I believe there are no exports to be added.

File: src/interpreter/ArithmeticHelper.ts [hunk 1]
```diff
@@ -47,7 +47,7 @@ export class ArithmeticHelper {
   private readonly actualEps: number
   constructor(
     private readonly config: Config,
-    private readonly dateTimeHelper: DateTimeHelper,
+    public readonly dateTimeHelper: DateTimeHelper,
     private readonly numberLiteralsHelper: NumberLiteralHelper,
```
Comment (wojciechczerniak) [inline src/interpreter/ArithmeticHelper.ts:51]: Why not make both public? If this is the recommended way to use them, they should be available

**Review by wojciechczerniak — CHANGES_REQUESTED** [review]
- This PR should add a section to docs but #741 is not yet merged. We have to wait

OMG, all those changes in visibility of interfaces 🙈 Dependency Injection could help

File: src/interpreter/plugin/BooleanPlugin.ts [hunk 2]
```diff
@@ -196,6 +198,18 @@ export class BooleanPlugin extends FunctionPlugin implements FunctionPluginTypec
...
+  //   if (ast.args.length === 0) {
+  //     return ArraySize.error()
+  //   }
+  //
+  //   const metadata = this.metadata('SWITCH')
+  //   const subChecks = ast.args.map((arg) => this.arraySizeForAst(arg, new InterpreterState(state.formulaAddress, state.arraysFlag || (metadata?.arrayFunction ?? false))))
+  //
+  //   const [{width, height}] = subChecks
+  //   return new ArraySize(width, height)
+  // }
```
[the remark this answers is not in the document]
Comment (izulin) [reply src/interpreter/plugin/BooleanPlugin.ts:211]: right

File: src/interpreter/Interpreter.ts [hunk 3]
```diff
@@ -43,22 +43,21 @@ import {FormulaVertex} from '../DependencyGraph/FormulaCellVertex'
...
     public readonly stats: Statistics,
-    public readonly dateHelper: DateTimeHelper,
-    public readonly numberLiteralsHelper: NumberLiteralHelper,
+    public readonly arithmeticHelper: ArithmeticHelper,
     public readonly functionRegistry: FunctionRegistry,
     public readonly namedExpressions: NamedExpressions,
-    public readonly serialization: Serialization
+    public readonly serialization: Serialization,
+    public readonly arraySizePredictor: ArraySizePredictor,
+    public readonly dateTimeHelper: DateTimeHelper
```
[the remark this answers is not in the document]
Comment (izulin) [reply src/interpreter/Interpreter.ts:58]: both classes are quite spacious in volume and separated in logic

[the remark this answers is not in the document]
Comment (wojciechczerniak) [reply src/interpreter/Interpreter.ts:58]: But we have access to `arithmeticHelper` that gives access to `dateTimeHelper` to `this.dateTimeHelper` isn't the same as `this.arithmeticHelper.dateTimeHelper` and we don't have to pass both args? If no then ok

[the remark this answers is not in the document]
Comment (izulin) [reply src/interpreter/Interpreter.ts:58]: I did a cleanup (in this PR), where objects explicitly used by major objects are passed explicitly, not extracted from other objects.
Helps when e.g. creating mocks where some args are `undefined`.

File: src/interpreter/plugin/MatrixPlugin.ts [hunk 4]
```diff
@@ -27,11 +27,26 @@ export interface KernelFunctionThis {
...
+function arraySizeForMultiplication(leftArraySize: ArraySize, rightArraySize: ArraySize): ArraySize {
+  return new ArraySize(rightArraySize.width, leftArraySize.height)
+}
+
+function arraySizeForPoolFunction(inputArray: ArraySize, windowSize: number, stride: number): ArraySize {
+  return new ArraySize(
+    1 + (inputArray.width - windowSize) / stride,
+    1 + (inputArray.height - windowSize) / stride,
+  )
+}
```
[the remark this answers is not in the document]
Comment (izulin) [reply src/interpreter/plugin/MatrixPlugin.ts]: inlined the first one, which is used once.
The others are useful, even if just for 2 or 3 usages in this plugin.

**Comment by izulin** [comment]
> OMG, all those changes in visibility of interfaces 🙈 Dependency Injection could help

All those interfaces should never be accessed by the end-user, and were made public only for internal usage.

**Comment by izulin** [comment]
> >Dependency Injection could help

Ok, I just educated myself on what Dependency Injection is, and AFAIR this is what I was striving for with the refactor.

File: src/interpreter/plugin/MatrixPlugin.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (wojciechczerniak) [reply src/interpreter/plugin/MatrixPlugin.ts]: I don't see more than 1 usage 🤔

**Review by wojciechczerniak — APPROVED** [review]
Thanks for addressing all my concerns. LGTM 👍

**Pull request merged**